### DISCUSSION TITLE
Can FastapiUsers be made compatible with CockroachDB?

### DISCUSSION THREAD

**Discussion by rbracco** [fastapi-users/fastapi-users#924] [Q&A]
I am interested in using CockroachDB as a cloud database, but it appears to be incompatible with pythons `databases` library. Specifically I am not able to pass a `databases.Database` instance to my `get_user_db` function. `get_user_db` calls `SQLAlchemyUserDatabase` which expects a `databases.Database` instance and makes calls on it like `fetch_one` which don't exist on the db Session generator that comes from my `get_db` function. Is there a workaround for getting a suitable database object to `SQLAlchemyUserDatabase` or are fastapi-users and CockroachDB fundamentally incompatible at this point in time? Thank you!

**Comment by frankie567** [comment]
Unless I'm mistaken, CockroachDB is designed to be 100% compatible with the PostgreSQL API, so there shouldn't be anything different than configuring a PostgreSQL database. Could you show me how you configure CockroachDB?

**Reply by kernfeld-cockroach** [reply]
> CockroachDB is designed to be 100% compatible with the PostgreSQL API, so there shouldn't be anything different than configuring a PostgreSQL database

CockroachDB is mostly Postgres-compatible, but not 100%. [Here](https://www.cockroachlabs.com/docs/stable/postgresql-compatibility.html) is a list of discrepancies.

**Comment by rbracco** [comment]
I think the issue is specifically within databases https://github.com/encode/databases/issues/369

Once I get things wired up and working I'll build my connection string and validate it with pydantic, but for now I just use a single string: `cockroachdb://user:password@instance...`. I am able to connect and read/write to CockroachDB via sqlalchemy but the issue is caused by the databases library as seen below. Any ideas?

![image](https://user-images.githubusercontent.com/47190785/154473951-b76c5063-ad6a-4849-8ec9-aab3db09da72.png)

![image](https://user-images.githubusercontent.com/47190785/154474371-ef41bbf5-1078-452c-b445-3c714f8348fd.png)

Inside `databases/core.py` here is the list of supported backends
![image](https://user-images.githubusercontent.com/47190785/154475221-d7914edc-2777-42e1-a41e-ce7903867a27.png)

**Reply by frankie567** [reply]
Yes, `databases` doesn't seem to have the required mapping with the `cockroachdb` scheme. Try to modify the scheme in your database URL by `postgresql`: `cockroachdb://user:password@instance...` becomes `postgresql://user:password@instance...`

**Reply by rbracco** [reply]
Hi, thank you for the quick response. Changing the databaseURL to be postgresql causes the following error
```
AssertionError: Could not determine version from string 'CockroachDB CCL v21.2.4-1-g70835279ac (x86_64-unknown-linux-gnu, built 2022/02/03 22:31:25, go1.16.6)'
```

But upon investigating further I found [this sqlalchemy-cockroachdb](https://github.com/cockroachdb/sqlalchemy-cockroachdb/issues/165) issue where someone notes you can just manually add cockroachdb as a supported backend, but point it at the postgres backend.  
![image](https://user-images.githubusercontent.com/47190785/154477531-dc9a4397-83dc-432f-9e17-10040752a8b2.png)

I altered my code to monkey-patch the supported backend and the app successfully loads, but then when I attempt to register a user I have the following error
![image](https://user-images.githubusercontent.com/47190785/154477792-31a6c565-e99d-45b6-88ec-454ff51ce652.png)
![image](https://user-images.githubusercontent.com/47190785/154479228-ff0f239c-e25c-4877-8eee-9491ee4f1a2b.png)

Update: I realized I needed to `await database.connect()` in `get_user_db()` and that causes 
![image](https://user-images.githubusercontent.com/47190785/154480650-cdb8cfe6-6219-413b-bb03-9256bf487186.png)

**Reply by frankie567** [reply]
Hmm, well, from what I understand of `sqlalchemy-cockroachdb`, it seems they have to implement a dedicated dialect to make it work. So, I don't think it'll work with `encode/databases` unless they specifically add support for it.

However, if you make the switch to `fastapi-users[sqlalchemy2]`, you'll have a pure SQLAlchemy configuration; so if you install `sqlalchemy-cockroachdb`, it should work.

**Reply by rbracco** [reply]
Thank you for the help! I'll look into those options.
